platform-sdk/docs/proposals/001-merkle-state-api/001-merkle-state-api.md

```diff
@@ -0,0 +1,65 @@
+# Merkle State API
+
+## Summary
+
+Refactor the Merkle tree so that it can be fully represented by the State API objects.
+Create a module that will provide a simple and easy-to-use API for the Merkle tree.
+
+## Purpose and Context
+
+### Why?
+
+The Merkle tree is a rather complex data structure, and to make interaction easier, we need to hide this complexity behind the API.
+Another reason to have this API is to provide additional flexibility in changing the implementation details of the Merkle tree.
+As long as they are hidden, we are free to change them.
+
+Note that this API is already implemented as a part of the Service Modularization project, but these classes will be moved
+to a separate module as a part of this proposal.
+
+### Context
+
+There are two issues that need to be addressed:
+
+* We need to have a public API for the Merkle tree that is simple and easy to use, residing in a separate module with a minimal set of dependencies.
+* The Merkle tree has to be fully expressed in terms of State API. Currently, `PlatformState` stands out and needs to be refactored.[](https://)
+
+## Requirements
+
+- `PlatformState` should be refactored to a singleton object as defined by the State API.
+- there should be a new module - `swirlds-state-api` - to host a set of interfaces, records, and abstract classes that represent the Hedera state.
+  This module should have a minimal set of dependencies. The Block Node should not have a compile-time dependency
+  on any other modules but this one to interact with the state.
+- there should be a new module - `swirlds-state-impl` module to host the implementation of these interfaces and helper classes.
+
+## Changes
+
+### Public API
+
+State API classes and interfaces will migrate to a designated module. The classes themselves will not change. `PlatformState` class will become a part of the public API. [](https://)
```

Comment:
The usage part is not clear for me. I assume that it may be useful for the Block Node, but I don't know how exactly it's going to be useful. 

For the internal use (in platform-sdk or services) there is an alternative interface that provides direct access to the `PlatformState` instance. 

To sum up - this platform state refactoring is driven by the fact that we need to be the Merkle tree uniformed and hide its complexity behind the scene, not by necessity of exposing `PlatofrmState` as a part of API

Reply:
> The usage part is not clear for me. I assume that it may be useful for the Block Node, but I don't know how exactly it's going to be useful.

Block nodes will read the block stream that contains state changes. It will then apply these state changes to its local state so that it has an up to date state from which to answer queries and produce state proofs.

> For the internal use (in platform-sdk or services) there is an alternative interface that provides direct access to the PlatformState instance.
I'm confused here. I thought platform will have to access `PlatformState` through the State API as well?

Reply:
> I thought platform will have to access PlatformState through the State API as well?

Even though it's possible to make it so, there is no benefit in a such a change. Why bother looking up a singleton by the service name and then get a value from there, if we can use the following interface instead? 

```
public interface MerkleRoot extends MerkleInternal {
    [skipped]
    /**
     * Get the platform state.
     *
     * @return the platform state
     */
    @NonNull
    PlatformState getPlatformState();
    /**
    [skipped]
}
````